lib/types.nix

```diff
@@ -246,7 +246,25 @@ rec {
     either = t1: t2: mkOptionType {
       name = "${t1.name} or ${t2.name}";
       check = x: t1.check x || t2.check x;
-      merge = mergeOneOption;
+      merge = loc: defs:
+             if all t1.check (getValues defs) then t1.merge loc defs
```

Comment:
And this would be fine, because submodules is a refined attribute set.  So this would still be ambiguous.  And this might be even worse with `either submodule attrs`, as all attribute sets would be considered as valid submodules containing undeclared options.